compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -199,14 +199,32 @@ impl<'a> Parser<'a> {
     ) -> PResult<'a, FnRetTy> {
         Ok(if self.eat(&token::RArrow) {
             // FIXME(Centril): Can we unconditionally `allow_plus`?
-            let ty = self.parse_ty_common(
-                allow_plus,
-                AllowCVariadic::No,
-                recover_qpath,
-                recover_return_sign,
-                None,
-                RecoverQuestionMark::Yes,
-            )?;
+
+            let mut tys = vec![];
+            let lo = self.token.span;
+            loop {
```

Comment:
The issue with this that we shouldn't recover this only for return types, we should be doing this further down, in `parse_ty_common`, maybe adding another argument to it to explicitly state what should be coming after it (to support things like `let foo: A | B = ...;`).